날짜 | 범위
--- | ---
1/23 | 10.2 커스텀 에러와 에러 확장 ~ 11.2 프라미스

### 커스텀 에러와 에러 확장
```javascript
class MyError extends Error {
  constructor(message){
    super(message);
    this.name = this.constructor.name;  //중복코드 줄이기
  }
}

class ValidationError extends MyError {
  constructor(message){
    super(message);
    //name을 지정할 필요가 없음
  }
}
```
- `instanceof`로 에러 유형을 확인하는 것을 권장한다.
- 주요 함수코드 내부에서 에러 처리를 하고 모든 에러를 포함할 수 있는 추상에러를 만들어 바깥으로 던지도록 한다. 해당 함수를 호출한 외부에서 추상 에러를 받고 `err.cause`같은 구체적인 에러 정보를 받아 확인하는 방식을 예외 감싸기라고 한다.

### 콜백
- 비동기적으로 수행하는 함수 다음에 실행되어야 하는 함수를 인수로 제공하는 방식을 콜백 기반 비동기 프로그래밍 이라고 한다.
```javascript
//오류 우선 콜백
test( function(error, success){
  if (error){
    //에러 처리
  }else{
    //성공 처리
  }
});
```
- 콜백 속의 콜백이 반복되면 깊은 중첩 코드가 만들어지는데 이를 콜백 지옥, 멸망의 피라미드라고 일컫는다.

### 프라미스
```javascript
let promise = new Promise(function(resolve, reject){    //excutor 함수

  //excutor코드 : new Promise가 만들어질때 자동으로 실행된다.
  if(success){
    resolve(value);  //성공
  }else{
    reject(error);  //에러
  }
});

promise.finally(alert('종료'));  //Promise가 처리되면 무조건 실행됨

promise.then(
  result => alert(result),  //성공했을때 실행되는 함수
  error => alert(error)  //실패했을때 실행되는 함수
);

promise.catch(alert);  //then(null, alert)와 동일함
```
- excutor함수의 인수 `resolve`, `reject`는 자바스크립트가 자체적으로 제공하는 콜백이다.
- excutor코드 내부에서 `resolve`나 `reject` 둘 중 하나는 반드시 호출해야 한다.
- `new Promise`가 반환하는 `promise`객체는 `state`와 `result` 프로퍼티를 갖는다.
- `resolve`나 `reject`가 호출된 이후에 호출되는 것은 무시된다.
- `resolve`와 `rejcet`는 인수를 하나만 받거나 혹은 아무것도 받지 않는다.
- `finally` 핸들러에서는 인수가 없기 때문에 `Promise`가 성공했는지 실패했는지 알 수 없다.
